### DISCUSSION TITLE
Using CMS Signer with different digest algorithms

### DISCUSSION THREAD

**Discussion by mholschbach** [Keyfactor/signserver-ce#51] [Q&A]
The CSM Signer from Signserver 6 uses shake256len as digest algorithm. Is there a configuration parameter to select other digest algorithms like shake256 (a different OID than shake256len) or sha256?

**Comment by netmackan** [comment]
Does not sound familiar.
What signature algorithm (SIGNATUREALGORITHM worker property) are you using?

**Reply by mholschbach** [reply]
I followed this tutorial https://doc.primekey.com/signserver/tutorials-and-guides/tutorial-signserver-post-quantum-signing to configure the CMS Signer. The SIGNATUREALGORITHM is set to Dilithium.

**Reply by netmackan** [reply]
Okey then I see:

Currently it is not configurable in the CMSSigner to use an other content digest algorithm then the for the signature default one. The default one is taken from DefaultDigestAlgorithmIdentifierFinder in Bouncy Castle (except for SPHINCS+ which is currently hard-coded to SHA-256 as it were not available in BC at the time of implementation https://github.com/Keyfactor/signserver-ce/blob/255fa3034702b8e1e62e6bfc960e2709af3a092d/signserver/modules/SignServer-Module-CMSSigner/src/main/java/org/signserver/module/cmssigner/CMSSigner.java#L339-L342).

In DefaultDigestAlgorithmIdentifierFinder the Dilitihium algorithms all use shake-256-len:
https://github.com/bcgit/bc-java/blob/9032821b326fba9fec7528899dfffa79f9f19aef/pkix/src/main/java/org/bouncycastle/operator/DefaultDigestAlgorithmIdentifierFinder.java#L228-L230

Potentially, one could consider implementing a new worker property say DIGESTALGORITHM and if specified use that with setContentDigest().
